### PR TITLE
Adding Configuration Interaction projects

### PR DESCRIPTION
**Pull request by tyzhang1993**
Hello,
Jeff Schriber and I are adding CI projects to psi4numpy!
Thank @dgasmith for the invitation!
Sam

 - [x] CIS
 - [x] CISD
 - [x] FCI

### CODE REVIEW

**Comment by jeffschriber** [comment]
I think we have a few more boxes to check before this PR is ready for review:
 - [ ] Implement Davidson-Liu solver
 - [ ] Add CI tutorial

**Comment by dgasmith** [comment]
Not super easy to do, but to get a CIWavefunction python side for manipulation you can do the following:


```
import numpy as np
import psi4

np.set_printoptions(precision=7, linewidth=200, threshold=2000, suppress=True)

# Set molecule to dimer
molecule {
    He 0.0 0.0 0.0
    He 0.0 0.0 1000.0
    symmetry c1
}


psi4.set_options({"BASIS": "6-31G",
                  })


# Build the SCF Wavefunction
scf_e, scf_wfn = psi4.energy("HF", return_wfn=True)

# Build integrals
mints = psi4.core.MintsHelper(scf_wfn.basisset())

# Build a CI Wavefunction
psi4.core.prepare_options_for_module("DETCI")
ciwfn = psi4.core.CIWavefunction(scf_wfn)
mints.integrals()
ciwfn.transform_ci_integrals()

# Compute the Hamiltonian
H = np.array(ciwfn.hamiltonian(0))
print(H)
```

**Comment by jeffschriber** [comment]
CIWavefunction is definitely the easiest/best way to do this, especially considering the Davidson-Liu solver. Maybe one consequence is that it does hide a lot of details, particularly with how determinants are generated. For example, the FCI code shows explicitly how to generate the determinant space,

    detList = []
    for alpha in combinations(xrange(nmo), ndocc):
        for beta in combinations(xrange(nmo), ndocc):
            detList.append(Determinant(alphaObtList=alpha, betaObtList=beta))

which is good purely for learning the theory, but does not really reflect how FCI algorithms are actually written. Would it be best to (a) just use CIWavefunction throughout, (b) have two versions (Current version for educational purposes, CIWavefunction with DL version for better representation of CI algorithms), or (c) move the current version into a CI tutorial only, and use CIWavefunction elsewhere? I'm leaning towards this last option, but I'm not too sure what is best.

**Comment by dgasmith** [comment]
I like c) a lot as well. However, it may require the most work. The tutorial side of things could likely be a team effort with myself and David helping.

BTW I thought I had it laying around, you can find a DL solver using the CIWave class [here](https://gist.github.com/dgasmith/3ca259921b3af8988eda5355cf80182f).

**Comment by loriab** [comment]
I think there's a problem with your history – you should have only your commits in the PR. You'll want to rebase to master in another branch then push to this PR's branch to heal.

```
(CI) >>> git checkout -b fixCI
(fixCI) >>> git pull --rebase upstream master
(fixCI) >>> git push -f fixCI:CI
```

**Comment by tyzhang1993** [comment]
Thank you! I hope I have solved the issue.

**Comment by dgasmith** [comment]
@tyzhang1993 Looking good!

**Comment by dgasmith** [comment]
Any updates on the DL solver or CI tutorial?

File: Configuration-Interaction/CIS.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+print('..finished transformation in %.3f seconds.\n' % (time.time() - t))
+
+from helper_CI import Determinant, HamiltonianGenerator
+from itertools import combinations
+
+print('Generating %d CIS singlet Determinants...' % (nDet_S + 1))
+t = time.time()
+
+occList = [i for i in xrange(ndocc)]
```
Comment (dgasmith) [inline Configuration-Interaction/CIS.py]: `xrange` will not work in Python 3, `range` is fine.

File: Configuration-Interaction/helper_CI.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+                
+        return dets
+    
+    def generateDoubleExcitationsOfDet(self, nmo):
+        """generate all the double excitations of determinant in a list"""
+        alphaO, betaO = self.getOrbitalIndexLists()
+        alphaU, betaU = self.getUnoccupiedOrbitalsInLists(nmo)
+        dets = []
+                
+        for i in alphaO:
```
Comment (dgasmith) [inline Configuration-Interaction/helper_CI.py]: Can you add a note that of what type of excitations each of these are?

File: Configuration-Interaction/helper_CI.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    def createFromIntTuple(intTuple):
+        return Determinant(alphaObtBits=intTuple[0], betaObtBits=intTuple[1])
+    
+    def generateSingleExcitationsOfDet(self, nmo):
+        """generate all the single excitations of determinant in a list"""
+        alphaO, betaO = self.getOrbitalIndexLists()
+        alphaU, betaU = self.getUnoccupiedOrbitalsInLists(nmo)
+        dets = []
+        
+        for i in alphaO:
```
Comment (dgasmith) [inline Configuration-Interaction/helper_CI.py]: Ditto as below.

File: Configuration-Interaction/helper_CI.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+__copyright__ = "(c) 2014-2017, The Psi4NumPy Developers"
+__license__   = "BSD-3-Clause"
+__date__      = "2017-05-26"
+
+from itertools import combinations
+
+class Determinant:
+    '''
+    class for determinant stored in bits
```
Comment (dgasmith) [inline Configuration-Interaction/helper_CI.py]: Can you add a bit more here about how the data bit layout works?

Comment (dgasmith) [inline Configuration-Interaction/helper_CI.py]: Might also want to detail (very briefly) how the bit operations work in python `^,&,etc`.

File: Configuration-Interaction/helper_CI.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        else:
+            return 0.0
+    
+    def calcMatrixElementDiffIn2(self, det1, det2):
+        """Calculate a matrix element by two determinants where the determinants differ by 2 spin orbitals"""
+        unique1, unique2, sign = det1.getUniqueOrbitalsInMixIndexListsPlusSign(det2)
+        return sign * self.antiSym2eInt[unique1[0], unique1[1], unique2[0], unique2[1]]
+    
+    def calcMatrixElementDiffIn1(self, det1, det2):
+        """Calculate a matrix element by two determinants where the determinants differ by 1 spin orbitals"""
```
Comment (dgasmith) [inline Configuration-Interaction/helper_CI.py]: For each of these could you write the matrix forms of the slater rules? `H[m, n] = sign * ( h_{m,n} + <m p || n p >)` or the like.

**Comment by dgasmith** [comment]
Also, please add the CI folder to `.scripts/travis_test.py:32` so that it will be included in the tests.

**Comment by jeffschriber** [comment]
Sorry, haven't been able to work on this for a few days. I've just about finished the CISD tutorial in addition to incorporating the DL solver for both it and FCI. Should be ready to go today.

**Comment by tyzhang1993** [comment]
@dgasmith Thank you very much for the comments!

File: Configuration-Interaction/CI_DL.py [hunk 6]
```diff
@@ -117,20 +112,14 @@
 for CI_ITER in range(max_guess - 1):
 
     # Subspace Matrix, Gij = < bi | H | bj >
-    #for i in range(num_vecs - num_eig, num_vecs):
     for i in range(0, num_vecs):
         # Build sigma for each b
         ciwfn.sigma(cvecs, svecs, i, i)
```
Comment (jeffschriber) [inline Configuration-Interaction/CI_DL.py]: @dgasmith @tyzhang1993 For some reason, the sigma vectors for roots other than the ground state are totally wrong on the first iteration. It does seem to give the right sigma vector for new subspace vectors added at each iteration. Not sure why the behavior is different at the first iteration.

Comment (dgasmith) [inline Configuration-Interaction/CI_DL.py]: Do `cvecs` and `svecs` have the right number of vectors to write into? You can also try a `cvecs.read(i, 0)` and `svecs.read(i, 0)` before hand. There might be some issue with it not paying attention to your `i` root.

Anything stand out about the resulting vector? Is it totally wrong, a different root, etc?

Comment (tyzhang1993) [inline Configuration-Interaction/CI_DL.py]: @dgasmith Thank you for the response, the sigma vectors for roots other than the ground states are totally wrong.

Comment (dgasmith) [inline Configuration-Interaction/CI_DL.py]: Did you try the `read` trick above and what results do you see?

Comment (jeffschriber) [inline Configuration-Interaction/CI_DL.py]: cvecs and svecs are both initialized with `max_guess` number of vectors, but the program exits when I put in a `svecs.read(i,0)`, even on the first iteration of `i` which gives the correct vector:

    PSIO_ERROR: Can't find TOC Entry buffer_ 0
    PSIO_ERROR: unit = 201, errval = 13
    PSIO_ERROR: 13 (no such TOC entry)
    Traceback (most recent call last):
      File "CI_DL.py", line 118, in <module>
        svecs.read(i,0)
    RuntimeError: 
    Fatal Error: PSIO Error

Comment (dgasmith) [inline Configuration-Interaction/CI_DL.py]: Ah, sorry you need to write before you can read. If you look up you can see that we write out blank lines to `dvecs`, you would need to do the same for `svecs`. Running the following snippet is fine:

```
H = np.array(ciwfn.hamiltonian(0))

# Setup test vecs
C1 = np.random.rand(H.shape[0])
C2 = np.random.rand(H.shape[0])

dvecs.np[:] = C1
dvecs.write(0, 1)
dvecs.symnormalize(1 / np.linalg.norm(C1), 0)
C1 = np.array(dvecs)

dvecs.np[:] = C2
dvecs.write(1, 0)
dvecs.symnormalize(1 / np.linalg.norm(C2), 1)
C2 = np.array(dvecs)

# Compute Explicit sigma
S1 = np.dot(H, C1)
S2 = np.dot(H, C2)
print(" ")

ciwfn.sigma(dvecs, svecs, 0, 0)
print(np.allclose(S1, svecs))

ciwfn.sigma(dvecs, svecs, 1, 1)
print(np.allclose(S2, svecs))
```

Is the excited state vector both normalized and symmetrized?

Comment (dgasmith) [inline Configuration-Interaction/CI_DL.py]: The `CIWave` module wasnt really built to do the above so a little patience with it is required.

**Comment by dgasmith** [comment]
Any updates on this? If you send me a code snippet that it behaving oddly I can take a look at it.

**Comment by dgasmith** [comment]
@jeffschriber @tyzhang1993 Hey guys, any update on this? I want to get another round of the paper out to everyone this week.

**Comment by jeffschriber** [comment]
Sorry for putting this off for so long, but I think I've just about solved the problem. I'm currently testing it with some excited states.

**Comment by jeffschriber** [comment]
Everything's working now (multiple roots, arbitrary number of guess vectors), I'll clean it up and push ASAP.

**Comment by dgasmith** [comment]
@jeffschriber Great, is everything ready to go?

**Comment by tyzhang1993** [comment]
Yes, it is ready to go. @dgasmith

**Pull request merged**